engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/InputMeta.scala

```diff
@@ -0,0 +1,102 @@
+  * @param headers - event headers converted to map
+  * @param leaderEpoch - number of leaders previously assigned by the controller (> 0 indicates leader failure)
+  * @tparam K - type of event key
+  */
+case class InputMeta[K](key: K,
+                        topic: String,
+                        partition: Integer,
+                        offset: java.lang.Long,
+                        timestamp: java.lang.Long,
+                        timestampType: TimestampType,
+                        headers: java.util.Map[String, String],
+                        leaderEpoch: Integer)
+
+object InputMeta {
+
+  val keyParameterName: String = "key"
+
+  /**
+    * Provides definition of whole metadata object, with given key type definition (keyTypingResult).
+    * See also [[InputMetaAwareTypeInformationCustomisation]]
+    */
+  def withType(keyTypingResult: typing.TypingResult): typing.TypingResult = {
+    // TODO: exclude non-key parameters to trait BaseKafkaInputMetaVariables and use it in TypesInformationExtractor.mandatoryClasses
+    new TypedObjectTypingResult(
+      ListMap(
+        keyParameterName -> keyTypingResult,
+        "topic" -> Typed[String],
+        "partition" -> Typed[Integer],
+        "offset" -> Typed[java.lang.Long],
+        "timestamp" -> Typed[java.lang.Long],
+        "timestampType" -> Typed[TimestampType],
+        "headers" -> TypedClass(classOf[java.util.Map[_, _]], List(Typed[String], Typed[String])),
+        "leaderEpoch" -> Typed[Integer]
+      ),
+      Typed.typedClass[InputMeta[AnyRef]]
+    ) {
+      override def display: String = Typed.genericTypeClass(classOf[InputMeta[_]], List(keyTypingResult)).display
+    }
+  }
+
+  def typeInformation[K](keyTypeInformation: TypeInformation[K]): CaseClassTypeInfo[InputMeta[K]] = {
+    val fieldNames = List(
+      keyParameterName,
+      "topic",
+      "partition",
+      "offset",
+      "timestamp",
+      "timestampType",
+      "headers",
+      "leaderEpoch"
+    )
+    val fieldTypes = List(
+      keyTypeInformation,
+      TypeInformation.of(classOf[String]),
+      TypeInformation.of(classOf[Integer]),
+      TypeInformation.of(classOf[java.lang.Long]),
+      TypeInformation.of(classOf[java.lang.Long]),
+      TypeInformation.of(classOf[TimestampType]),
+      new MapTypeInfo(classOf[String], classOf[String]),
+      TypeInformation.of(classOf[Integer])
+    )
+    new CaseClassTypeInfo[InputMeta[K]](classOf[InputMeta[K]], Array.empty, fieldTypes, fieldNames){
+      override def createSerializer(config: ExecutionConfig): TypeSerializer[InputMeta[K]] =
+        new ScalaCaseClassSerializer[InputMeta[K]](classOf[InputMeta[K]], fieldTypes.map(_.createSerializer(config)).toArray)
+    }
+  }
+}
+
+/**
+  * Customisation for TypeInformationDetection that provides type information for BaseInputMeta.
+  */
+class InputMetaAwareTypeInformationCustomisation extends TypingResultAwareTypeInformationCustomisation {
```

Comment:
Shouldn't it be used in generic/demo/dev model?

Reply:
:+1: Currently it does not seem to be used anywhere? At least we should register it with ServiceLoader?

Reply:
Fixed. See `META-INF/services` in `flinkKafkaUtil` and `process`. Example test scenario is in `InputMetaDeserializationSpec`.